tools/doc/json.js

```diff
@@ -272,21 +272,27 @@ function parseSignature(text, sig) {
   var params = text.match(paramExpr);
   if (!params) return;
   params = params[1];
-  // the [ is irrelevant. ] indicates optionalness.
-  params = params.replace(/\[/g, '');
   params = params.split(/,/);
+  var optionalLevel = 0;
   params.forEach(function(p, i, _) {
     p = p.trim();
     if (!p) return;
     var param = sig.params[i];
     var optional = false;
     var def;
+
     // [foo] -> optional
-    if (p.charAt(p.length - 1) === ']') {
-      optional = true;
-      p = p.replace(/\]/g, '');
-      p = p.trim();
+    var character;
+    while ((character = p.charAt(0)) === '[' || character === ']') {
```

Comment:
I don't really like this form of assigning and checking two things in the same part of the loop (also below). Wouldn't this be easier as a:

```
for(const character of p) {
    if(character === ']' || character === '[') break;
    optionalLevel += {'[': 1, ']': -1}[character];
}
```